plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/index/HudiRecordLevelIndexSupport.java

```diff
@@ -48,81 +57,87 @@ public class HudiRecordLevelIndexSupport
 
     public static final String DEFAULT_COLUMN_VALUE_SEPARATOR = ":";
     public static final String DEFAULT_RECORD_KEY_PARTS_SEPARATOR = ",";
+    private final CompletableFuture<Optional<Set<String>>> relevantFileIdsFuture;
+    private final Duration recordIndexWaitTimeout;
+    private final long futureStartTimeMs;
 
-    public HudiRecordLevelIndexSupport(SchemaTableName schemaTableName, Lazy<HoodieTableMetaClient> lazyMetaClient)
+    public HudiRecordLevelIndexSupport(ConnectorSession session, SchemaTableName schemaTableName, Lazy<HoodieTableMetaClient> lazyMetaClient, Lazy<HoodieTableMetadata> lazyTableMetadata, TupleDomain<HiveColumnHandle> regularColumnPredicates)
     {
         super(log, schemaTableName, lazyMetaClient);
-    }
-
-    @Override
-    public Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(
-            HoodieTableMetadata metadataTable,
-            Map<String, List<FileSlice>> inputFileSlices,
-            TupleDomain<String> regularColumnPredicates)
-    {
-        // Should not happen since canApply checks for this, include for safety
+        this.recordIndexWaitTimeout = getRecordIndexWaitTimeout(session);
         if (regularColumnPredicates.isAll()) {
             log.debug("Predicates cover all data, skipping record level index lookup.");
-            return inputFileSlices;
+            this.relevantFileIdsFuture = CompletableFuture.completedFuture(Optional.empty());
         }
+        else {
+            this.relevantFileIdsFuture = CompletableFuture.supplyAsync(() -> {
```

Comment:
Sth to take care of in a follow-up: if the query is cancelled or the query has finished, we should stop the record index and secondary index lookup to avoid resource occupation on the coordination.